plugins/by-name/trouble/default.nix

```diff
@@ -115,200 +115,274 @@ helpers.neovim-plugin.mkNeovimPlugin {
-      mapAttrs
-        (
-          diagnostic: default:
-          helpers.defaultNullOpts.mkStr default "Icon/text for ${diagnostic} diagnostics."
-        )
+    modes =
+      defaultNullOpts.mkAttrsOf lib.types.anything
         {
-          error = "";
-          warning = "";
-          hint = "";
-          information = "";
-          other = "﫠";
-        };
+          lsp_references = {
+            params = {
+              include_declaration = true;
+            };
+          };
+          lsp_base = {
+            params = {
+              include_current = false;
+            };
+          };
+          symbols = {
+            desc = "document symbols";
+            mode = "lsp_document_symbols";
+            focus = false;
+            win = {
+              position = "right";
+            };
+            filter = {
+              "not" = {
+                ft = "lua";
+                kind = "Package";
+              };
+              any = {
+                ft = [
+                  "help"
+                  "markdown"
+                ];
+                kind = [
+                  "Class"
+                  "Constructor"
+                  "Enum"
+                  "Field"
+                  "Function"
+                  "Interface"
+                  "Method"
+                  "Module"
+                  "Namespace"
+                  "Package"
+                  "Property"
+                  "Struct"
+                  "Trait"
+                ];
+              };
+            };
+          };
+        }
+        ''
+          Customize the various builtin modes or define your own.
+        '';
 
-    use_diagnostic_signs = helpers.defaultNullOpts.mkBool false ''
-      Enabling this will use the signs defined in your lsp client
-    '';
+    icons = defaultNullOpts.mkAttrsOf lib.types.anything {
+      indent = {
+        top = "│ ";
+        middle = "├╴";
+        last = "└╴";
+        fold_open = " ";
+        fold_closed = " ";
+        ws = "  ";
+      };
+      folder_closed = " ";
+      folder_open = " ";
+      kinds = {
+        Array = " ";
+        Boolean = "󰨙 ";
+        Class = " ";
+        Constant = "󰏿 ";
+        Constructor = " ";
+        Enum = " ";
+        EnumMember = " ";
+        Event = " ";
+        Field = " ";
+        File = " ";
+        Function = "󰊕 ";
+        Interface = " ";
+        Key = " ";
+        Method = "󰊕 ";
+        Module = " ";
+        Namespace = "󰦮 ";
+        Null = " ";
+        Number = "󰎠 ";
+        Object = " ";
+        Operator = " ";
+        Package = " ";
+        Property = " ";
+        String = " ";
+        Struct = "󰆼 ";
+        TypeParameter = " ";
+        Variable = "󰀫 ";
+      };
+    } "Define custom icons.";
   };
 
   extraConfig = cfg: {
+    # TODO: introduced 2024-10-12: remove after 24.11
+    warnings =
+      let
+        definedOpts = lib.filter (opt: lib.hasAttrByPath (lib.toList opt) cfg.settings) [
```

Comment:
This is more general/future-proof, and maybe we'll end up factoring it out into `lib.deprecation` if it becomes a common pattern.

But if you wanted to, we can simplify here because none of _these_ are nested attr paths:

```nix
definedOpts = lib.filter (opt: cfg.settings ? ${opt}) [
```

or
```nix
definedOpts = lib.filter (lib.flip lib.hasAttr cfg.settings) [
```